File: libraries/typescript/src/providers/deepgram-flux-stt.ts
/**
 * Deepgram Flux Streaming STT Adapter for Cloudflare Workers AI & Deepgram Flux WebSocket endpoint.
 *
 * Connects via WebSocket to streaming `@cf/deepgram/flux` or custom Cloudflare proxy endpoints,
 * parsing `TurnInfo` messages and mapping `StartOfTurn`, `Update`, `EagerEndOfTurn`,
 * `TurnResumed`, and `EndOfTurn` events into Patter-normalized {@link Transcript} objects.
 */

import WebSocket from 'ws';
import type { IncomingMessage } from 'http';
import { AuthenticationError, PatterConnectionError, RateLimitError } from '../errors';
import { getLogger } from '../logger';
import type { Transcript, DeepgramWord } from './deepgram-stt';

export type FluxEventType = 'StartOfTurn' | 'Update' | 'EagerEndOfTurn' | 'TurnResumed' | 'EndOfTurn';

export interface FluxTurnInfoMessage {
  readonly type: 'TurnInfo' | 'Metadata';
  readonly request_id?: string;
  readonly sequence_id?: number;
  readonly event?: FluxEventType;
  readonly turn_index?: number;
  readonly audio_window_start?: number;
  readonly audio_window_end?: number;
  readonly transcript?: string;
  readonly words?: ReadonlyArray<DeepgramWord>;
  readonly end_of_turn_confidence?: number;
}

export interface DeepgramFluxSTTOptions {
  readonly apiKey?: string;
  readonly url?: string; // WebSocket URL (e.g. wss://api.deepgram.com/v1/listen or Cloudflare Worker URL)
  readonly model?: string; // Default "@cf/deepgram/flux" or "flux-general-en"
  readonly language?: string; // Default "en"
  readonly sampleRate?: number; // Default 16000 Hz
  readonly encoding?: string; // Default "linear16"
}

type TranscriptCallback = (transcript: Transcript & { fluxEvent?: FluxEventType; endOfTurnConfidence?: number }) => void;
type ErrorCallback = (error: Error) => void;

export class DeepgramFluxSTT {
  static readonly providerKey = 'deepgram_flux';
  private ws: WebSocket | null = null;
  private readonly transcriptCallbacks = new Set<TranscriptCallback>();
  private readonly errorCallbacks = new Set<ErrorCallback>();
  private running = false;
  private requestId = '';
  private readonly ctorOptions: DeepgramFluxSTTOptions;

  get isRunning(): boolean {
    return this.running;
  }

  constructor(options: DeepgramFluxSTTOptions = {}) {
    this.ctorOptions = options;
  }

  clone(): this {
    const ctor = this.constructor as new (opts: DeepgramFluxSTTOptions) => this;
    return new ctor(this.ctorOptions);
  }

  async connect(): Promise<void> {
    const url = this.ctorOptions.url ?? 'wss://api.deepgram.com/v1/listen?model=flux-general-en&encoding=linear16&sample_rate=16000';
    const apiKey = this.ctorOptions.apiKey ?? process.env.DEEPGRAM_API_KEY ?? '';

    const headers: Record<string, string> = {};
    if (apiKey) {
      headers.Authorization = `Token ${apiKey}`;
    }

    const ws = new WebSocket(url, { headers });
    this.ws = ws;

    await new Promise<void>((resolve, reject) => {
      let settled = false;
      const timer = setTimeout(() => {
        if (!settled) {
          settled = true;
          reject(new PatterConnectionError('Deepgram Flux WebSocket connect timeout'));
        }
      }, 10000);

      ws.once('open', () => {
        if (!settled) {
          settled = true;
          clearTimeout(timer);
          resolve();
        }
      });

      ws.once('error', (err: Error) => {
        if (!settled) {
          settled = true;
          clearTimeout(timer);
          reject(err);
        }
      });

      ws.once('unexpected-response', (_req: unknown, res: IncomingMessage) => {
        const status = res?.statusCode ?? 0;
        if (!settled) {
          settled = true;
          clearTimeout(timer);
          if (status === 401 || status === 403) {
            reject(new AuthenticationError(`Deepgram Flux rejected API key (HTTP ${status})`));
          } else if (status === 429) {
            reject(new RateLimitError(`Deepgram Flux rate limit exceeded (HTTP ${status})`));
          } else {
            reject(new PatterConnectionError(`Deepgram Flux WebSocket upgrade failed (HTTP ${status})`));
          }
        }
      });
    });

    this.running = true;
    ws.on('message', (data) => this.handleMessage(data.toString()));
    ws.on('close', () => {
      this.running = false;
    });
    ws.on('error', (err) => {
      this.emitError(err);
    });
  }

  handleMessage(raw: string): void {
    let msg: FluxTurnInfoMessage;
    try {
      msg = JSON.parse(raw) as FluxTurnInfoMessage;
    } catch {
      return;
    }

    if (msg.type === 'Metadata' && msg.request_id) {
      this.requestId = msg.request_id;
      return;
    }

    if (msg.type !== 'TurnInfo' && !msg.event) return;

    const event = msg.event;
    const text = (msg.transcript ?? '').trim();
    const requestId = msg.request_id ?? this.requestId;
    const endOfTurnConfidence = msg.end_of_turn_confidence ?? 0;

    if (event === 'StartOfTurn') {
      this.emitTranscript({
        text: '',
        isFinal: false,
        confidence: 0,
        eventType: 'SpeechStarted',
        requestId,
        fluxEvent: 'StartOfTurn',
      });
      return;
    }

    if (event === 'Update') {
      this.emitTranscript({
        text,
        isFinal: false,
        confidence: 0.8,
        eventType: 'Results',
        words: msg.words,
        requestId,
        fluxEvent: 'Update',
      });
      return;
    }

    if (event === 'EagerEndOfTurn') {
      this.emitTranscript({
        text,
        isFinal: true,
        speechFinal: true,
        confidence: endOfTurnConfidence || 0.75,
        eventType: 'Results',
        words: msg.words,
        requestId,
        fluxEvent: 'EagerEndOfTurn',
        endOfTurnConfidence,
      });
      return;
    }

    if (event === 'TurnResumed') {
      this.emitTranscript({
        text,
        isFinal: false,
        confidence: 0.5,
        eventType: 'SpeechStarted',
        requestId,
        fluxEvent: 'TurnResumed',
      });
      return;
    }

    if (event === 'EndOfTurn') {
      this.emitTranscript({
        text,
        isFinal: true,
        speechFinal: true,
        confidence: endOfTurnConfidence || 0.95,
        eventType: 'UtteranceEnd',
        words: msg.words,
        requestId,
        fluxEvent: 'EndOfTurn',
        endOfTurnConfidence,
      });
    }
  }

  sendAudio(audio: Buffer): void {
    if (this.ws && this.ws.readyState === WebSocket.OPEN && audio.length > 0) {
      this.ws.send(audio);
    }
  }

  onTranscript(callback: TranscriptCallback): void {
    this.transcriptCallbacks.add(callback);
  }

  offTranscript(callback: TranscriptCallback): void {
    this.transcriptCallbacks.delete(callback);
  }

  onError(callback: ErrorCallback): void {
    this.errorCallbacks.add(callback);
  }

  offError(callback: ErrorCallback): void {
    this.errorCallbacks.delete(callback);
  }

  finalize(): void {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      try {
        this.ws.send(JSON.stringify({ type: 'Finalize' }));
      } catch {
        // ignore
      }
    }
  }

  close(): void {
    this.running = false;
    if (this.ws) {
      try {
        this.ws.send(JSON.stringify({ type: 'CloseStream' }));
        this.ws.close();
      } catch {
        // ignore
      }
      this.ws = null;
    }
  }

  private emitTranscript(transcript: Transcript & { fluxEvent?: FluxEventType; endOfTurnConfidence?: number }): void {
    for (const cb of this.transcriptCallbacks) {
      try {
        Promise.resolve(cb(transcript)).catch((err) =>
          getLogger().error(`DeepgramFluxSTT transcript callback failed: ${String(err)}`),
        );
      } catch (err) {
        getLogger().error(`DeepgramFluxSTT transcript callback threw: ${String(err)}`);
      }
    }
  }

  private emitError(err: Error): void {
    for (const cb of this.errorCallbacks) {
      try {
        cb(err);
      } catch (cbErr) {
        getLogger().error(`DeepgramFluxSTT error callback threw: ${String(cbErr)}`);
      }
    }
  }
}

File: libraries/python/getpatter/providers/deepgram_flux_stt.py
"""
Deepgram Flux Streaming STT Adapter for Cloudflare Workers AI & Deepgram Flux WebSocket endpoint.

Connects via WebSocket to streaming `@cf/deepgram/flux` or custom Cloudflare proxy endpoints,
parsing `TurnInfo` messages and mapping `StartOfTurn`, `Update`, `EagerEndOfTurn`,
`TurnResumed`, and `EndOfTurn` events into Patter-normalized Transcript objects.
"""

from __future__ import annotations

import asyncio
import json
import logging
from typing import AsyncIterator, ClassVar, Optional, Union
from urllib.parse import urlencode

import websockets
from websockets.exceptions import InvalidStatus

from getpatter.exceptions import (
    AuthenticationError,
    PatterConnectionError,
    RateLimitError,
)
from getpatter.providers.base import STTProvider, Transcript

logger = logging.getLogger("getpatter.providers.deepgram_flux_stt")

DEFAULT_FLUX_WS_URL = "wss://api.deepgram.com/v1/listen?model=flux-general-en&encoding=linear16&sample_rate=16000"


class DeepgramFluxSTT(STTProvider):
    provider_key: ClassVar[str] = "deepgram_flux"

    def __init__(
        self,
        api_key: str = "",
        url: str = "",
        model: str = "@cf/deepgram/flux",
        language: str = "en",
        sample_rate: int = 16000,
        encoding: str = "linear16",
    ) -> None:
        self.api_key = api_key
        self.url = url
        self.model = model
        self.language = language
        self.sample_rate = sample_rate
        self.encoding = encoding
        self._ws: Optional[websockets.WebSocketClientProtocol] = None
        self.request_id: Optional[str] = None
        self._audio_bytes_sent = 0

    async def connect(self) -> None:
        ws_url = self.url or DEFAULT_FLUX_WS_URL
        headers = {}
        if self.api_key:
            headers["Authorization"] = f"Token {self.api_key}"

        try:
            self._ws = await asyncio.wait_for(
                websockets.connect(ws_url, additional_headers=headers),
                timeout=10.0,
            )
        except InvalidStatus as exc:
            status_code = getattr(getattr(exc, "response", None), "status_code", None)
            if status_code in (401, 403):
                raise AuthenticationError(
                    f"Deepgram Flux rejected API key (HTTP {status_code})."
                ) from exc
            if status_code == 429:
                raise RateLimitError("Deepgram Flux rate limit exceeded (HTTP 429).") from exc
            raise PatterConnectionError(
                f"Deepgram Flux WebSocket upgrade failed (HTTP {status_code})."
            ) from exc
        except (OSError, asyncio.TimeoutError) as exc:
            raise PatterConnectionError(
                f"Failed to connect to Deepgram Flux: {exc}"
            ) from exc

    async def send_audio(self, audio_chunk: bytes) -> None:
        if self._ws is None:
            raise RuntimeError("Not connected. Call connect() first.")
        if not audio_chunk:
            return
        self._audio_bytes_sent += len(audio_chunk)
        await self._ws.send(audio_chunk)

    async def finalize(self) -> None:
        if self._ws is not None:
            try:
                await self._ws.send(json.dumps({"type": "Finalize"}))
            except Exception:
                pass

    def _parse_message(self, raw_message: str) -> Optional[Transcript]:
        try:
            data = json.loads(raw_message)
        except Exception:
            return None

        msg_type = data.get("type", "")
        if msg_type == "Metadata":
            self.request_id = data.get("request_id")
            return None

        if msg_type != "TurnInfo" and "event" not in data:
            return None

        event = data.get("event")
        text = (data.get("transcript") or "").strip()
        request_id = data.get("request_id") or self.request_id
        end_of_turn_confidence = float(data.get("end_of_turn_confidence", 0.0))
        words = data.get("words", []) or []

        if event == "StartOfTurn":
            return Transcript(
                text="",
                is_final=False,
                confidence=0.0,
                event_type="SpeechStarted",
                request_id=request_id,
            )

        if event == "Update":
            return Transcript(
                text=text,
                is_final=False,
                confidence=0.8,
                event_type="Results",
                words=tuple(words),
                request_id=request_id,
            )

        if event == "EagerEndOfTurn":
            return Transcript(
                text=text,
                is_final=True,
                speech_final=True,
                confidence=end_of_turn_confidence or 0.75,
                event_type="Results",
                words=tuple(words),
                request_id=request_id,
            )

        if event == "TurnResumed":
            return Transcript(
                text=text,
                is_final=False,
                confidence=0.5,
                event_type="SpeechStarted",
                request_id=request_id,
            )

        if event == "EndOfTurn":
            return Transcript(
                text=text,
                is_final=True,
                speech_final=True,
                confidence=end_of_turn_confidence or 0.95,
                event_type="UtteranceEnd",
                words=tuple(words),
                request_id=request_id,
            )

        return None

    async def receive_transcripts(self) -> AsyncIterator[Transcript]:
        if self._ws is None:
            raise RuntimeError("Not connected. Call connect() first.")

        async for raw_message in self._ws:
            if isinstance(raw_message, bytes):
                continue
            transcript = self._parse_message(raw_message)
            if transcript is not None:
                yield transcript

    async def close(self) -> None:
        if self._ws is not None:
            try:
                await self._ws.send(json.dumps({"type": "CloseStream"}))
                await self._ws.close()
            except Exception:
                pass
            self._ws = None
